src/org/jitsi/impl/neomedia/rtp/translator/SimulcastController.java

```diff
@@ -157,102 +134,118 @@ public long getTargetSSRC()
+        if (sourceTL0Idx > -1)
+        {
+            sourceTL0Idx
+                = sourceEncodings[sourceTL0Idx].getBaseLayer().getIndex();
+        }
 
+        boolean sourceTL0IsActive = sourceEncodings[sourceTL0Idx].isActive();
         if (!sourceFrameDesc.isIndependent()
-            || !sourceFrameDesc.getRTPEncoding().isActive()
-            || sourceIdx == transformState.currentIdx)
+            || !sourceTL0IsActive
+            || sourceTL0Idx == currentTL0Idx)
         {
             // An intra-codec switch requires a key frame.
-            boolean accept = sourceSSRC == transformState.currentSSRC;
 
-            if (accept)
+            long sourceSSRC = sourceFrameDesc.getRTPEncoding().getPrimarySSRC();
+            boolean accept = sourceSSRC == bitstreamController.getTL0SSRC();
+
+            if (!accept)
             {
-                onAccept(buf, off, len);
+                return false;
             }
 
-            return accept;
+            // Give the bitstream filter a chance to drop the packet.
+            return bitstreamController.accept(sourceFrameDesc, buf, off, len);
         }
 
-        if ((targetIdx <= sourceIdx && sourceIdx < transformState.currentIdx)
-            || (transformState.currentIdx < sourceIdx && sourceIdx <= targetIdx)
-            || (!currentRTPEncodingIsActive && sourceIdx <= targetIdx))
+        if ((targetTL0Idx <= sourceTL0Idx && sourceTL0Idx < currentTL0Idx)
+            || (currentTL0Idx < sourceTL0Idx && sourceTL0Idx <= targetTL0Idx)
+            || (!currentTL0IsActive && sourceTL0Idx <= targetTL0Idx))
         {
-            // Pretend this is the next frame of whatever has already been
-            // sent.
+            long maxTs = bitstreamController.getMaxTs(),
+                transmittedBytes = bitstreamController.getTransmittedBytes(),
+                transmittedPackets = bitstreamController.getTransmittedPackets();
 
-            long tsDelta; int seqNumDelta;
-            if (filterState.maxSeqNum != -1)
-            {
-                seqNumDelta = (filterState.maxSeqNum
-                    + 1 - sourceFrameDesc.getStart()) & 0xFFFF;
-                tsDelta = (filterState.maxTs
-                    + 3000 - sourceFrameDesc.getTimestamp()) & 0xFFFFFFFFL;
-            }
-            else
-            {
-                seqNumDelta = 0; tsDelta = 0;
-            }
+            int maxSeqNum = bitstreamController.getMaxSeqNum();
 
-            if (logger.isInfoEnabled())
+            synchronized (this)
             {
-                logger.info("new_transform src_ssrc=" + sourceSSRC
-                    + ",src_idx=" + sourceIdx
-                    + ",ts_delta=" + tsDelta
-                    + ",seq_delta=" + seqNumDelta);
+                int targetIdx = bitstreamController.getTargetIndex(),
+                    optimalIdx = bitstreamController.getOptimalIndex();
+
+                bitstreamController = new BitstreamController(
```

Comment:
Have you thought about passing the old BitstreamController as a parameter to the constructor of the new one? It looks to me that BitstreamController itself should know best which fields it needs to inherit, and it can do so without exposing fields such as transmittedBytes, maxTs, etc.